clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
I feel a little more inclined to try and make this API work since the implementation feels more natural, but I agree we should make sure we understand the impact and revert if we can't find a nice way to deal with it. Since `pollOnce` is an internal API, I'm not sure it actually has to invoke poll. At one point, I considered renaming it to `pollRecords` instead to emphasize that the main point was fetching data. If we did that, then I would see no inconsistency in shortcutting when records are already available. 

I think the thing that we're having a little trouble wrapping our heads around is the fact that a single fetch response can affect two invocations of `fetchedRecords`. Basically we're trying to deal with this by introducing another call to fetchedRecords. Another way we could handle it is to create a new method, for example, `Fetcher.outOfRangePartitions`, which would allow us to move the exception throw into KafkaConsumer and remove the need for the extra call to fetchedRecords. What do you think?